### PR TITLE
Visitor based saved games

### PR DESCRIPTION
**Pull request by MicroVirus**
Use visitors for the saved game code, continuing with what was done earlier for some but by far not all classes, rather than memcpy-ing the class data.
This is a huge step towards supporting a different saved game format (such as XML or JSON).

This change does almost certainly break the crc8 functionality until also the crc8 can be written in terms of visitors.

Also fixed an issue with firm_die_array not being properly saved/loaded from file.

Next up is converting the code in OGFILE2.cpp that determines the order of visiting to the visitor pattern.

### CODE REVIEW

**Comment by MicroVirus** [comment]
I'll have to test and find out what the underlying cause is when running under Linux, and if the saved games were and are exchangeable between Windows and Linux. I'll try and get to this the coming week.

**Comment by the3dfxdude** [comment]
I think the Version1 saves aren't working too. A large number of scenarios fail with "Game load error".
4 10 11 13 14 15 16 17 18 19 20 23 24 25 26 27 28 29 30 31 36 37 38

All tutorials fail.

**Comment by the3dfxdude** [comment]
I am going to close this pull request for now. I fixed a few of the issues, but then got stuck on an issue in the DynArray* changes. The upcoming release is priority now.

File: src/file_writer.cpp [hunk 1]
```diff
@@ -74,15 +75,18 @@ bool FileWriter::skip(size_t len)
...
 {
    if (!this->ok)
       return false;
 
    if (this->original_type != File::STRUCTURED)
       return true;
 
-   return this->write<uint16_t>(size);
+   if (static_cast<unsigned int>(size) > std::numeric_limits<uint16_t>::max())
+      return this->write<uint16_t>(0);
```
[the remark this answers is not in the document]
Comment (MicroVirus) [reply src/file_writer.cpp]: No, this is the way the original `File::file_write` class writes: add the record size as an unsigned short if <=0xFFFF, else 0.

**Pull request closed without merging**